// twitch-chat-client/es/index.mjs
/* eslint-disable filenames/match-exported */
import { deprecateClass } from '@d-fischer/shared-utils';
import { ChatClient } from "./ChatClient.mjs";
/** @deprecated Use the named export `ChatClient` instead. */
var DeprecatedChatClient = deprecateClass(ChatClient, "[twitch-chat-client] The default export has been deprecated. Use the named export instead:\n\n\timport { ChatClient } from 'twitch-chat-client';");
/** @deprecated Use the named export `ChatClient` instead. */
export default DeprecatedChatClient;
export { ChatClient };
export { TwitchPrivateMessage as PrivateMessage } from "./StandardCommands/TwitchPrivateMessage.mjs";
export { ClearChat } from "./Capabilities/TwitchCommandsCapability/MessageTypes/ClearChat.mjs";
export { HostTarget } from "./Capabilities/TwitchCommandsCapability/MessageTypes/HostTarget.mjs";
export { Reconnect } from "./Capabilities/TwitchCommandsCapability/MessageTypes/Reconnect.mjs";
export { RoomState } from "./Capabilities/TwitchCommandsCapability/MessageTypes/RoomState.mjs";
export { UserNotice } from "./Capabilities/TwitchCommandsCapability/MessageTypes/UserNotice.mjs";
export { UserState } from "./Capabilities/TwitchCommandsCapability/MessageTypes/UserState.mjs";
export { Whisper } from "./Capabilities/TwitchCommandsCapability/MessageTypes/Whisper.mjs";
export { ClearMsg } from "./Capabilities/TwitchTagsCapability/MessageTypes/ClearMsg.mjs";
export { GlobalUserState } from "./Capabilities/TwitchTagsCapability/MessageTypes/GlobalUserState.mjs";
export { ChatUser } from "./ChatUser.mjs";
export { LogLevel } from '@d-fischer/logger';
export { toChannelName, toUserName } from "./Toolkit/UserTools.mjs";
