// twitch/es/index.mjs
/* eslint-disable filenames/match-exported */
import { deprecateClass } from '@d-fischer/shared-utils';
import { ApiClient } from "./ApiClient.mjs";
/** @deprecated Use the named export `ApiClient` instead. */
var DeprecatedTwitchClient = deprecateClass(ApiClient, "[twitch] The default export has been deprecated. Use the named export instead:\n\n\timport { ApiClient } from 'twitch';");
/** @deprecated Use the named export `ApiClient` instead. */
export default DeprecatedTwitchClient;
export { ApiClient };
export { ChatBadgeList } from "./API/Badges/ChatBadgeList.mjs";
export { ChatBadgeSet } from "./API/Badges/ChatBadgeSet.mjs";
export { ChatBadgeVersion } from "./API/Badges/ChatBadgeVersion.mjs";
export { HelixPaginatedRequest } from "./API/Helix/HelixPaginatedRequest.mjs";
export { HelixBitsLeaderboard } from "./API/Helix/Bits/HelixBitsLeaderboard.mjs";
export { HelixBitsLeaderboardEntry } from "./API/Helix/Bits/HelixBitsLeaderboardEntry.mjs";
export { HelixCheermoteList } from "./API/Helix/Bits/HelixCheermoteList.mjs";
export { HelixChannel } from "./API/Helix/Channel/HelixChannel.mjs";
export { HelixCustomReward } from "./API/Helix/ChannelPoints/HelixCustomReward.mjs";
export { HelixCustomRewardRedemption } from "./API/Helix/ChannelPoints/HelixCustomRewardRedemption.mjs";
export { HelixClip } from "./API/Helix/Clip/HelixClip.mjs";
export { HelixExtensionTransaction } from "./API/Helix/Extensions/HelixExtensionTransaction.mjs";
export { HelixGame } from "./API/Helix/Game/HelixGame.mjs";
export { HelixHypeTrainContribution } from "./API/Helix/HypeTrain/HelixHypeTrainContribution.mjs";
export { HelixHypeTrainEvent } from "./API/Helix/HypeTrain/HelixHypeTrainEvent.mjs";
export { HelixBan } from "./API/Helix/Moderation/HelixBan.mjs";
export { HelixBanEvent } from "./API/Helix/Moderation/HelixBanEvent.mjs";
export { HelixModerator } from "./API/Helix/Moderation/HelixModerator.mjs";
export { HelixModeratorEvent } from "./API/Helix/Moderation/HelixModeratorEvent.mjs";
export { HelixChannelSearchResult } from "./API/Helix/Search/HelixChannelSearchResult.mjs";
export { HelixStream, HelixStreamType } from "./API/Helix/Stream/HelixStream.mjs";
export { HelixStreamMarker } from "./API/Helix/Stream/HelixStreamMarker.mjs";
export { HelixStreamMarkerWithVideo } from "./API/Helix/Stream/HelixStreamMarkerWithVideo.mjs";
export { HelixSubscription } from "./API/Helix/Subscriptions/HelixSubscription.mjs";
export { HelixSubscriptionEvent } from "./API/Helix/Subscriptions/HelixSubscriptionEvent.mjs";
export { HelixTag } from "./API/Helix/Tag/HelixTag.mjs";
export { HelixFollow } from "./API/Helix/User/HelixFollow.mjs";
export { HelixPrivilegedUser } from "./API/Helix/User/HelixPrivilegedUser.mjs";
export { HelixBroadcasterType, HelixUser, HelixUserType } from "./API/Helix/User/HelixUser.mjs";
export { HelixVideo } from "./API/Helix/Video/HelixVideo.mjs";
export { HelixWebHookSubscription } from "./API/Helix/WebHooks/HelixWebHookSubscription.mjs";
export { Subscription } from "./API/Kraken/Subscription.mjs";
export { CheermoteList } from "./API/Kraken/Bits/CheermoteList.mjs";
export { Channel } from "./API/Kraken/Channel/Channel.mjs";
export { ChannelFollow } from "./API/Kraken/Channel/ChannelFollow.mjs";
export { ChannelPlaceholder } from "./API/Kraken/Channel/ChannelPlaceholder.mjs";
export { ChannelSubscription } from "./API/Kraken/Channel/ChannelSubscription.mjs";
export { EmoteSetList } from "./API/Kraken/Channel/EmoteSetList.mjs";
export { PrivilegedChannel } from "./API/Kraken/Channel/PrivilegedChannel.mjs";
export { ChatEmote } from "./API/Kraken/Chat/ChatEmote.mjs";
export { ChatEmoteList } from "./API/Kraken/Chat/ChatEmoteList.mjs";
export { ChatRoom } from "./API/Kraken/Chat/ChatRoom.mjs";
export { Stream, StreamType } from "./API/Kraken/Stream/Stream.mjs";
export { Team } from "./API/Kraken/Team/Team.mjs";
export { TeamWithUsers } from "./API/Kraken/Team/TeamWithUsers.mjs";
export { PrivilegedUser } from "./API/Kraken/User/PrivilegedUser.mjs";
export { User } from "./API/Kraken/User/User.mjs";
export { UserBlock } from "./API/Kraken/User/UserBlock.mjs";
export { UserChatInfo } from "./API/Kraken/User/UserChatInfo.mjs";
export { UserFollow } from "./API/Kraken/User/UserFollow.mjs";
export { UserSubscription } from "./API/Kraken/User/UserSubscription.mjs";
export { CreatedVideo } from "./API/Kraken/Video/CreatedVideo.mjs";
export { Video } from "./API/Kraken/Video/Video.mjs";
export { ChattersList } from "./API/Unsupported/ChattersList.mjs";
export { CheermoteBackground, CheermoteScale, CheermoteState } from "./API/Shared/BaseCheermoteList.mjs";
export { ConfigError } from "./Errors/ConfigError.mjs";
export { HellFreezesOverError } from "./Errors/HellFreezesOverError.mjs";
export { InvalidTokenTypeError } from "./Errors/InvalidTokenTypeError.mjs";
export { NoSubscriptionProgramError } from "./Errors/NoSubscriptionProgramError.mjs";
export { StreamNotLiveError } from "./Errors/StreamNotLiveError.mjs";
export { extractUserId, extractUserName } from "./Toolkit/UserTools.mjs";
/** @deprecated Import this from `twitch-api-call` instead. */
export { HttpStatusCodeError, TwitchApiCallType } from 'twitch-api-call';
/** @deprecated Import `HttpStatusCodeError` instead. */
export { HttpStatusCodeError as HTTPStatusCodeError } from 'twitch-api-call';
/** @deprecated Import `TwitchApiCallType` instead. */
export { TwitchApiCallType as TwitchAPICallType } from 'twitch-api-call';
/** @deprecated Import this from `twitch-auth` instead. */
export { AccessToken, InvalidTokenError, StaticAuthProvider, ClientCredentialsAuthProvider, RefreshableAuthProvider, TokenInfo } from 'twitch-auth';
