// dank-twitch-irc/dist/message/parser/tag-values.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tagParserFor = exports.convertToFlags = exports.convertToEmoteSets = exports.convertToEmotes = exports.convertToBadges = exports.convertToTimestamp = exports.convertToColor = exports.convertToBoolean = exports.convertToInt = exports.convertToString = exports.getData = exports.requireData = void 0;
const badges_1 = require("./badges");
const color_1 = require("./color");
const emote_sets_1 = require("./emote-sets");
const emotes_1 = require("./emotes");
const flags_1 = require("./flags");
const missing_tag_error_1 = require("./missing-tag-error");
const parse_error_1 = require("./parse-error");
function requireData(ircTags, key, converter, ...converterArgs) {
    const stringValue = ircTags[key];
    if (stringValue == null) {
        throw new missing_tag_error_1.MissingTagError(key, stringValue);
    }
    const value = converter(stringValue, ...converterArgs);
    if (value == null) {
        throw new missing_tag_error_1.MissingTagError(key, stringValue);
    }
    return value;
}
exports.requireData = requireData;
function getData(ircTags, key, converter, ...converterArgs) {
    const stringValue = ircTags[key];
    if (stringValue == null) {
        return undefined;
    }
    return converter(stringValue, ...converterArgs);
}
exports.getData = getData;
function convertToString(value) {
    return value;
}
exports.convertToString = convertToString;
function convertToInt(value) {
    const parsedInt = parseInt(value);
    if (isNaN(parsedInt)) {
        throw new parse_error_1.ParseError(`Failed to parse integer from tag value "${value}"`);
    }
    return parsedInt;
}
exports.convertToInt = convertToInt;
function convertToBoolean(value) {
    return Boolean(convertToInt(value));
}
exports.convertToBoolean = convertToBoolean;
function convertToColor(value) {
    if (value.length <= 0) {
        return undefined;
    }
    return color_1.parseColor(value);
}
exports.convertToColor = convertToColor;
function convertToTimestamp(value) {
    return new Date(convertToInt(value));
}
exports.convertToTimestamp = convertToTimestamp;
function convertToBadges(value) {
    return badges_1.parseBadges(value);
}
exports.convertToBadges = convertToBadges;
function convertToEmotes(value, messageText) {
    return emotes_1.parseEmotes(messageText, value);
}
exports.convertToEmotes = convertToEmotes;
function convertToEmoteSets(value) {
    return emote_sets_1.parseEmoteSets(value);
}
exports.convertToEmoteSets = convertToEmoteSets;
function convertToFlags(value, messageText) {
    return flags_1.parseFlags(messageText, value);
}
exports.convertToFlags = convertToFlags;
function tagParserFor(ircTags) {
    return {
        getString: (key) => getData(ircTags, key, convertToString),
        requireString: (key) => requireData(ircTags, key, convertToString),
        getInt: (key) => getData(ircTags, key, convertToInt),
        requireInt: (key) => requireData(ircTags, key, convertToInt),
        getBoolean: (key) => getData(ircTags, key, convertToBoolean),
        requireBoolean: (key) => requireData(ircTags, key, convertToBoolean),
        getColor: (key) => getData(ircTags, key, convertToColor),
        requireColor: (key) => requireData(ircTags, key, convertToColor),
        getTimestamp: (key) => getData(ircTags, key, convertToTimestamp),
        requireTimestamp: (key) => requireData(ircTags, key, convertToTimestamp),
        getBadges: (key) => getData(ircTags, key, convertToBadges),
        requireBadges: (key) => requireData(ircTags, key, convertToBadges),
        getEmotes: (key, messageText) => getData(ircTags, key, convertToEmotes, messageText),
        requireEmotes: (key, messageText) => requireData(ircTags, key, convertToEmotes, messageText),
        getEmoteSets: (key) => getData(ircTags, key, convertToEmoteSets),
        requireEmoteSets: (key) => requireData(ircTags, key, convertToEmoteSets),
        getFlags: (key, messageText) => getData(ircTags, key, convertToFlags, messageText),
    };
}
exports.tagParserFor = tagParserFor;
