// dank-twitch-irc/dist/operations/get-mods-vips.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVips = exports.getMods = exports.GetUsersError = void 0;
const await_response_1 = require("../await/await-response");
const conditions_1 = require("../await/conditions");
const errors_1 = require("../client/errors");
const getModsInfo = {
    command: "mods",
    msgIdError: "usage_mods",
    msgIdNone: "no_mods",
    msgIdSome: "room_mods",
    someMessagePrefix: "The moderators of this channel are: ",
    someMessageSuffix: "",
};
const getVipsInfo = {
    command: "vips",
    msgIdError: "usage_vips",
    msgIdNone: "no_vips",
    msgIdSome: "vips_success",
    someMessagePrefix: "The VIPs of this channel are: ",
    someMessageSuffix: ".",
};
class GetUsersError extends errors_1.MessageError {
    constructor(channelName, type, message, cause) {
        super(message, cause);
        this.channelName = channelName;
        this.type = type;
    }
}
exports.GetUsersError = GetUsersError;
async function getMods(conn, channelName) {
    return await getModsOrVips(conn, channelName, getModsInfo);
}
exports.getMods = getMods;
async function getVips(conn, channelName) {
    return await getModsOrVips(conn, channelName, getVipsInfo);
}
exports.getVips = getVips;
async function getModsOrVips(conn, channelName, config) {
    conn.sendRaw(`PRIVMSG #${channelName} :/${config.command}`);
    const responseMsg = (await await_response_1.awaitResponse(conn, {
        success: conditions_1.matchingNotice(channelName, [config.msgIdNone, config.msgIdSome]),
        failure: conditions_1.matchingNotice(channelName, [config.msgIdError]),
        errorType: (msg, cause) => new GetUsersError(channelName, config.command, msg, cause),
        errorMessage: `Failed to get ${config.command} of channel ${channelName}`,
    }));
    if (responseMsg.messageID === config.msgIdNone) {
        return [];
    }
    if (responseMsg.messageID === config.msgIdSome) {
        let text = responseMsg.messageText;
        if (!text.startsWith(config.someMessagePrefix) ||
            !text.endsWith(config.someMessageSuffix)) {
            throw new GetUsersError(channelName, config.command, `Failed to get ${config.command} of channel ${channelName}: Response message had unexpected format: ${responseMsg.rawSource}`);
        }
        // slice away the prefix and suffix
        text = text.slice(config.someMessagePrefix.length, text.length - config.someMessageSuffix.length);
        return text.split(", ");
    }
    throw new Error("unreachable");
}
