// dank-twitch-irc/dist/operations/say.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.me = exports.say = exports.SayError = exports.removeCommands = void 0;
const await_response_1 = require("../await/await-response");
const errors_1 = require("../client/errors");
const notice_1 = require("../message/twitch-types/notice");
const userstate_1 = require("../message/twitch-types/userstate");
const privmsg_1 = require("./privmsg");
function removeCommands(message) {
    if (message.startsWith(".") || message.startsWith("/")) {
        return `/ ${message}`;
    }
    else {
        return message;
    }
}
exports.removeCommands = removeCommands;
class SayError extends errors_1.MessageError {
    constructor(failedChannelName, failedMessage, action, message, cause) {
        super(message, cause);
        this.failedChannelName = failedChannelName;
        this.messageText = failedMessage;
        this.action = action;
    }
}
exports.SayError = SayError;
const badNoticeIDs = [
    "msg_banned",
    "msg_bad_characters",
    // If you believe this is an error, please rephrase and try again.
    "msg_channel_blocked",
    "msg_channel_suspended",
    "msg_duplicate",
    // less than 30 seconds ago.
    "msg_emoteonly",
    // the smiley in the chat text area.
    "msg_facebook",
    // your Twitch settings under the connections tab.
    "msg_followersonly",
    "msg_followersonly_followed",
    // <duration2>. Continue following to chat!
    "msg_followersonly_zero",
    "msg_r9k",
    "msg_ratelimit",
    "msg_rejected",
    "msg_rejected_mandatory",
    "msg_room_not_found",
    "msg_slowmode",
    // talk again in <number> seconds.
    "msg_subsonly",
    // https://www.twitch.tv/products/<broadcaster login name>/ticket?ref=subscriber_only_mode_chat.
    "msg_suspended",
    "msg_timedout",
    "msg_verified_email",
];
async function say(conn, channelName, messageText, action = false) {
    let command;
    let errorMessage;
    let errorType;
    if (action) {
        command = `/me ${messageText}`;
        errorMessage = `Failed to say [#${channelName}]: /me ${messageText}`;
        errorType = (msg, cause) => new SayError(channelName, messageText, true, msg, cause);
    }
    else {
        command = removeCommands(messageText);
        errorMessage = `Failed to say [#${channelName}]: ${messageText}`;
        errorType = (msg, cause) => new SayError(channelName, messageText, false, msg, cause);
    }
    privmsg_1.sendPrivmsg(conn, channelName, command);
    return await_response_1.awaitResponse(conn, {
        success: (msg) => msg instanceof userstate_1.UserstateMessage && msg.channelName === channelName,
        failure: (msg) => msg instanceof notice_1.NoticeMessage &&
            msg.channelName === channelName &&
            badNoticeIDs.includes(msg.messageID),
        errorType,
        errorMessage,
    });
}
exports.say = say;
async function me(conn, channelName, message) {
    return say(conn, channelName, message, true);
}
exports.me = me;
