// dank-twitch-irc/dist/operations/set-color.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setColor = exports.SetColorError = void 0;
const await_response_1 = require("../await/await-response");
const errors_1 = require("../client/errors");
const color_1 = require("../message/color");
const notice_1 = require("../message/twitch-types/notice");
const privmsg_1 = require("./privmsg");
class SetColorError extends errors_1.MessageError {
    constructor(wantedColor, message, cause) {
        super(message, cause);
        this.wantedColor = wantedColor;
    }
}
exports.SetColorError = SetColorError;
const badNoticeIDs = [
    "turbo_only_color",
    "usage_color",
];
async function setColor(conn, color) {
    const colorAsHex = color_1.colorToHexString(color);
    privmsg_1.sendPrivmsg(conn, conn.configuration.username, `/color ${colorAsHex}`);
    await await_response_1.awaitResponse(conn, {
        failure: (msg) => msg instanceof notice_1.NoticeMessage &&
            msg.channelName === conn.configuration.username &&
            badNoticeIDs.includes(msg.messageID),
        success: (msg) => msg instanceof notice_1.NoticeMessage &&
            msg.channelName === conn.configuration.username &&
            msg.messageID === "color_changed",
        errorType: (msg, cause) => new SetColorError(color, msg, cause),
        errorMessage: `Failed to set color to ${colorAsHex}`,
    });
}
exports.setColor = setColor;
