// dank-twitch-irc/dist/operations/timeout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.timeout = exports.UserTimeoutError = void 0;
const ms = require("ms");
const await_response_1 = require("../await/await-response");
const conditions_1 = require("../await/conditions");
const errors_1 = require("../client/errors");
const channel_1 = require("../validation/channel");
const privmsg_1 = require("./privmsg");
class UserTimeoutError extends errors_1.MessageError {
    constructor(channelName, username, length, reason, message, cause) {
        super(message, cause);
        this.channelName = channelName;
        this.username = username;
        this.length = length;
        this.reason = reason;
    }
}
exports.UserTimeoutError = UserTimeoutError;
const failureNoticeIDs = [
    "no_permission",
    "bad_timeout_admin",
    "bad_timeout_anon",
    "bad_timeout_broadcaster",
    "bad_timeout_duration",
    "bad_timeout_global_mod",
    "bad_timeout_mod",
    "bad_timeout_self",
    "bad_timeout_staff",
    "usage_timeout",
];
const successNoticeIDs = ["timeout_success", "already_banned"];
async function timeout(conn, channelName, username, length, reason) {
    channel_1.validateChannelName(channelName);
    channel_1.validateChannelName(username);
    let cmd;
    if (reason != null) {
        cmd = `/timeout ${username} ${length} ${reason}`;
    }
    else {
        cmd = `/timeout ${username} ${length}`;
    }
    await privmsg_1.sendPrivmsg(conn, channelName, cmd);
    await await_response_1.awaitResponse(conn, {
        success: conditions_1.matchingNotice(channelName, successNoticeIDs),
        failure: conditions_1.matchingNotice(channelName, failureNoticeIDs),
        errorType: (msg, cause) => new UserTimeoutError(channelName, username, length, reason, msg, cause),
        errorMessage: `Failed to timeout ${username} for ` +
            `${ms(length * 1000)} in #${channelName}`,
    });
}
exports.timeout = timeout;
