// dank-twitch-irc/dist/message/twitch-types/privmsg.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const badge_1 = require("../badge");
const badges_1 = require("../badges");
const twitch_message_1 = require("../parser/twitch-message");
const privmsg_1 = require("./privmsg");
describe("./message/twitch-types/privmsg", function () {
    describe("#parseActionAndMessage()", function () {
        it("should return non-actions unmodified", function () {
            chai_1.assert.deepStrictEqual(privmsg_1.parseActionAndMessage("HeyGuys"), {
                isAction: false,
                message: "HeyGuys",
            });
            chai_1.assert.deepStrictEqual(privmsg_1.parseActionAndMessage("\u0001ACTION HeyGuys"), {
                isAction: false,
                message: "\u0001ACTION HeyGuys",
            });
            chai_1.assert.deepStrictEqual(privmsg_1.parseActionAndMessage("HeyGuys\u0001"), {
                isAction: false,
                message: "HeyGuys\u0001",
            });
            // missing space
            chai_1.assert.deepStrictEqual(privmsg_1.parseActionAndMessage("\u0001ACTIONHeyGuys\u0001"), {
                isAction: false,
                message: "\u0001ACTIONHeyGuys\u0001",
            });
        });
        it("should remove action prefix and suffix on valid actions", function () {
            chai_1.assert.deepStrictEqual(privmsg_1.parseActionAndMessage("\u0001ACTION HeyGuys\u0001"), {
                isAction: true,
                message: "HeyGuys",
            });
            // nested
            chai_1.assert.deepStrictEqual(privmsg_1.parseActionAndMessage("\u0001ACTION \u0001ACTION HeyGuys\u0001\u0001"), {
                isAction: true,
                message: "\u0001ACTION HeyGuys\u0001",
            });
        });
    });
    describe("PrivmsgMessage", function () {
        it("should be able to parse a real PRIVMSG message", function () {
            const msgText = "@badge-info=subscriber/5;badges=broadcaster/1,subscriber/0;" +
                "color=#19E6E6;display-name=randers;emotes=;flags=;id=7eb848c9-1060-4e5e-9f4c-612877982e79;" +
                "mod=0;room-id=40286300;subscriber=1;tmi-sent-ts=1563096499780;turbo=0;" +
                "user-id=40286300;user-type= :randers!randers@randers.tmi.twitch.tv PRIVMSG #randers :test";
            const msg = twitch_message_1.parseTwitchMessage(msgText);
            chai_1.assert.instanceOf(msg, privmsg_1.PrivmsgMessage);
            chai_1.assert.strictEqual(msg.channelName, "randers");
            chai_1.assert.strictEqual(msg.messageText, "test");
            chai_1.assert.isFalse(msg.isAction);
            chai_1.assert.strictEqual(msg.senderUsername, "randers");
            chai_1.assert.strictEqual(msg.senderUserID, "40286300");
            chai_1.assert.deepStrictEqual(msg.badgeInfo, new badges_1.TwitchBadgesList(new badge_1.TwitchBadge("subscriber", "5")));
            chai_1.assert.strictEqual(msg.badgeInfoRaw, "subscriber/5");
            chai_1.assert.deepStrictEqual(msg.badges, new badges_1.TwitchBadgesList(new badge_1.TwitchBadge("broadcaster", "1"), new badge_1.TwitchBadge("subscriber", "0")));
            chai_1.assert.strictEqual(msg.badgesRaw, "broadcaster/1,subscriber/0");
            chai_1.assert.isUndefined(msg.bits);
            chai_1.assert.isUndefined(msg.bitsRaw);
            chai_1.assert.deepStrictEqual(msg.color, { r: 0x19, g: 0xe6, b: 0xe6 });
            chai_1.assert.strictEqual(msg.colorRaw, "#19E6E6");
            chai_1.assert.strictEqual(msg.displayName, "randers");
            chai_1.assert.deepStrictEqual(msg.emotes, []);
            chai_1.assert.strictEqual(msg.emotesRaw, "");
            chai_1.assert.strictEqual(msg.messageID, "7eb848c9-1060-4e5e-9f4c-612877982e79");
            chai_1.assert.isFalse(msg.isMod);
            chai_1.assert.strictEqual(msg.isModRaw, "0");
            chai_1.assert.strictEqual(msg.channelID, "40286300");
            chai_1.assert.strictEqual(msg.serverTimestamp.getTime(), 1563096499780);
            chai_1.assert.strictEqual(msg.serverTimestampRaw, "1563096499780");
            chai_1.assert.deepStrictEqual(msg.extractUserState(), {
                badgeInfo: new badges_1.TwitchBadgesList(new badge_1.TwitchBadge("subscriber", "5")),
                badgeInfoRaw: "subscriber/5",
                badges: new badges_1.TwitchBadgesList(new badge_1.TwitchBadge("broadcaster", "1"), new badge_1.TwitchBadge("subscriber", "0")),
                badgesRaw: "broadcaster/1,subscriber/0",
                color: { r: 0x19, g: 0xe6, b: 0xe6 },
                colorRaw: "#19E6E6",
                displayName: "randers",
                isMod: false,
                isModRaw: "0",
            });
            chai_1.assert.isFalse(msg.isCheer());
        });
        it("trims spaces at the end of display names", function () {
            const msgText = "@badge-info=subscriber/5;badges=broadcaster/1,subscriber/0;" +
                "color=#19E6E6;display-name=randers\\s;emotes=;flags=;id=7eb848c9-1060-4e5e-9f4c-612877982e79;" +
                "mod=0;room-id=40286300;subscriber=1;tmi-sent-ts=1563096499780;turbo=0;" +
                "user-id=40286300;user-type= :randers!randers@randers.tmi.twitch.tv PRIVMSG #randers :test";
            const msg = twitch_message_1.parseTwitchMessage(msgText);
            chai_1.assert.strictEqual(msg.displayName, "randers");
            chai_1.assert.strictEqual(msg.extractUserState().displayName, "randers");
        });
    });
});
