// dank-twitch-irc/dist/mixins/ignore-promise-rejections.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IgnoreUnhandledPromiseRejectionsMixin = void 0;
const apply_function_replacements_1 = require("../utils/apply-function-replacements");
const ignore_errors_1 = require("../utils/ignore-errors");
class IgnoreUnhandledPromiseRejectionsMixin {
    applyToClient(client) {
        const genericReplacement = (originalFn, ...args) => {
            const originalPromise = originalFn(...args);
            originalPromise.catch(ignore_errors_1.ignoreErrors);
            return originalPromise;
        };
        apply_function_replacements_1.applyReplacements(this, client, {
            join: genericReplacement,
            part: genericReplacement,
            privmsg: genericReplacement,
            say: genericReplacement,
            me: genericReplacement,
            whisper: genericReplacement,
            setColor: genericReplacement,
            ping: genericReplacement,
        });
    }
}
exports.IgnoreUnhandledPromiseRejectionsMixin = IgnoreUnhandledPromiseRejectionsMixin;
