// twitch-chat-client/es/StandardCommands/TwitchPrivateMessage.mjs
import { __extends, __read, __spread } from "tslib";
import { MessageTypes } from 'ircv3';
import { ChatUser } from "../ChatUser.mjs";
import { fillTextPositions, parseEmoteOffsets, parseEmotePositions } from "../Toolkit/EmoteTools.mjs";
var TwitchPrivateMessage = /** @class */ (function (_super) {
    __extends(TwitchPrivateMessage, _super);
    function TwitchPrivateMessage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(TwitchPrivateMessage.prototype, "userInfo", {
        get: function () {
            return new ChatUser(this._prefix.nick, this._tags);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TwitchPrivateMessage.prototype, "channelId", {
        get: function () {
            var _a;
            return (_a = this._tags.get('room-id')) !== null && _a !== void 0 ? _a : null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TwitchPrivateMessage.prototype, "isCheer", {
        get: function () {
            var _a;
            return (_a = this._tags.has('bits')) !== null && _a !== void 0 ? _a : false;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TwitchPrivateMessage.prototype, "totalBits", {
        get: function () {
            var _a;
            return Number((_a = this._tags.get('bits')) !== null && _a !== void 0 ? _a : 0);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TwitchPrivateMessage.prototype, "emoteOffsets", {
        get: function () {
            return parseEmoteOffsets(this._tags.get('emotes'));
        },
        enumerable: false,
        configurable: true
    });
    TwitchPrivateMessage.prototype.parseEmotes = function () {
        var messageText = this.params.message;
        var foundEmotes = parseEmotePositions(messageText, this.emoteOffsets);
        return fillTextPositions(messageText, foundEmotes);
    };
    TwitchPrivateMessage.prototype.parseEmotesAndBits = function (cheermotes) {
        var messageText = this.params.message;
        var foundCheermotes = cheermotes.parseMessage(messageText);
        var foundEmotesAndCheermotes = __spread(parseEmotePositions(messageText, this.emoteOffsets), foundCheermotes.map(function (cheermote) { return ({
            type: 'cheer',
            position: cheermote.position,
            length: cheermote.length,
            name: cheermote.name,
            amount: cheermote.amount,
            displayInfo: cheermote.displayInfo
        }); }));
        foundEmotesAndCheermotes.sort(function (a, b) { return a.position - b.position; });
        return fillTextPositions(messageText, foundEmotesAndCheermotes);
    };
    return TwitchPrivateMessage;
}(MessageTypes.Commands.PrivateMessage));
export { TwitchPrivateMessage };
