// twitch-api-call/es/apiCall.mjs
/// <reference lib="dom" />
import { __awaiter, __generator } from "tslib";
import fetch, { Headers } from '@d-fischer/cross-fetch';
import { stringify } from '@d-fischer/qs';
import { transformTwitchApiResponse } from "./helpers/transform.mjs";
import { getTwitchApiUrl } from "./helpers/url.mjs";
import { TwitchApiCallType } from "./TwitchApiCallOptions.mjs";
/**
 * Makes a call to the Twitch API using the given credentials, returning the raw Response object.
 *
 * @param options
 * @param clientId
 * @param accessToken
 */
export function callTwitchApiRaw(options, clientId, accessToken) {
    return __awaiter(this, void 0, void 0, function () {
        var type, url, params, headers, body, requestOptions;
        return __generator(this, function (_a) {
            type = options.type === undefined ? TwitchApiCallType.Kraken : options.type;
            url = getTwitchApiUrl(options.url, type);
            params = stringify(options.query, { arrayFormat: 'repeat' });
            headers = new Headers({
                Accept: type === TwitchApiCallType.Kraken
                    ? "application/vnd.twitchtv.v" + (options.version || 5) + "+json"
                    : 'application/json'
            });
            if (options.body) {
                body = stringify(options.body);
                headers.append('Content-Type', 'application/x-www-form-urlencoded');
            }
            else if (options.jsonBody) {
                body = JSON.stringify(options.jsonBody);
                headers.append('Content-Type', 'application/json');
            }
            if (clientId && type !== TwitchApiCallType.Auth) {
                headers.append('Client-ID', clientId);
            }
            if (accessToken) {
                headers.append('Authorization', (type === TwitchApiCallType.Helix ? 'Bearer' : 'OAuth') + " " + accessToken);
            }
            requestOptions = {
                method: options.method || 'GET',
                headers: headers,
                body: body
            };
            return [2 /*return*/, fetch(params ? url + "?" + params : url, requestOptions)];
        });
    });
}
/**
 * Makes a call to the Twitch API using given credentials.
 *
 * @param options The configuration of the call.
 * @param clientId The client ID of your application.
 * @param accessToken The access token to call the API with.
 *
 * You need to obtain one using one of the [Twitch OAuth flows](https://dev.twitch.tv/docs/authentication/getting-tokens-oauth/).
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function callTwitchApi(options, clientId, accessToken) {
    return __awaiter(this, void 0, void 0, function () {
        var response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, callTwitchApiRaw(options, clientId, accessToken)];
                case 1:
                    response = _a.sent();
                    return [2 /*return*/, transformTwitchApiResponse(response)];
            }
        });
    });
}
