// twitch-chat-client/lib/index.js
var __tsu = {
    cache: [],
    defineExport: function (name, def) {
        Object.defineProperty(exports, name, def);
        this.cache.push({ name: name, def: def });
    },
    redefineExports: function () {
        this.cache.forEach(function (exp) {
            Object.defineProperty(exports, exp.name, exp.def);
        });
    }
};
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toUserName = exports.toChannelName = exports.LogLevel = exports.ChatUser = exports.GlobalUserState = exports.ClearMsg = exports.Whisper = exports.UserState = exports.UserNotice = exports.RoomState = exports.Reconnect = exports.HostTarget = exports.ClearChat = exports.PrivateMessage = exports.ChatClient = void 0;
/* eslint-disable filenames/match-exported */
var shared_utils_1 = require("@d-fischer/shared-utils");
var ChatClient_1 = require("./ChatClient");
__tsu.defineExport("ChatClient", { enumerable: true, get: function () { return ChatClient_1.ChatClient; } });
/** @deprecated Use the named export `ChatClient` instead. */
var DeprecatedChatClient = shared_utils_1.deprecateClass(ChatClient_1.ChatClient, "[twitch-chat-client] The default export has been deprecated. Use the named export instead:\n\n\timport { ChatClient } from 'twitch-chat-client';");
/** @deprecated Use the named export `ChatClient` instead. */
exports.default = DeprecatedChatClient;
var TwitchPrivateMessage_1 = require("./StandardCommands/TwitchPrivateMessage");
__tsu.defineExport("PrivateMessage", { enumerable: true, get: function () { return TwitchPrivateMessage_1.TwitchPrivateMessage; } });
var ClearChat_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/ClearChat");
__tsu.defineExport("ClearChat", { enumerable: true, get: function () { return ClearChat_1.ClearChat; } });
var HostTarget_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/HostTarget");
__tsu.defineExport("HostTarget", { enumerable: true, get: function () { return HostTarget_1.HostTarget; } });
var Reconnect_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/Reconnect");
__tsu.defineExport("Reconnect", { enumerable: true, get: function () { return Reconnect_1.Reconnect; } });
var RoomState_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/RoomState");
__tsu.defineExport("RoomState", { enumerable: true, get: function () { return RoomState_1.RoomState; } });
var UserNotice_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/UserNotice");
__tsu.defineExport("UserNotice", { enumerable: true, get: function () { return UserNotice_1.UserNotice; } });
var UserState_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/UserState");
__tsu.defineExport("UserState", { enumerable: true, get: function () { return UserState_1.UserState; } });
var Whisper_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/Whisper");
__tsu.defineExport("Whisper", { enumerable: true, get: function () { return Whisper_1.Whisper; } });
var ClearMsg_1 = require("./Capabilities/TwitchTagsCapability/MessageTypes/ClearMsg");
__tsu.defineExport("ClearMsg", { enumerable: true, get: function () { return ClearMsg_1.ClearMsg; } });
var GlobalUserState_1 = require("./Capabilities/TwitchTagsCapability/MessageTypes/GlobalUserState");
__tsu.defineExport("GlobalUserState", { enumerable: true, get: function () { return GlobalUserState_1.GlobalUserState; } });
var ChatUser_1 = require("./ChatUser");
__tsu.defineExport("ChatUser", { enumerable: true, get: function () { return ChatUser_1.ChatUser; } });
var logger_1 = require("@d-fischer/logger");
__tsu.defineExport("LogLevel", { enumerable: true, get: function () { return logger_1.LogLevel; } });
var UserTools_1 = require("./Toolkit/UserTools");
__tsu.defineExport("toChannelName", { enumerable: true, get: function () { return UserTools_1.toChannelName; } });
__tsu.defineExport("toUserName", { enumerable: true, get: function () { return UserTools_1.toUserName; } });
module.exports = exports = DeprecatedChatClient;
Object.defineProperty(exports, "__esModule", { value: true });
__tsu.redefineExports();
/** @deprecated Use the named export `ChatClient` instead. */
exports.default = DeprecatedChatClient;
